### PR TITLE
CASM-3555, CASM-3556: Handle concurrent and finer-grained updates

### PR DESCRIPTION
**Pull request by ekamin-hpe**
CASM-3555, CASM-3556: Handle concurrent and finer-grained updates

This commit adds a fix to provide better concurrency support when
updating the config map. When patching the config map, specify the
resource_version metadata field. This results in an HTTP 409 (conflict)
error if the target resource_version has unexpectedly changed (i.e.
some other process updated the config map first). This allows the process
to only update the config map when something else has not first updated it.

In order to provide finer-grained updates, introduce a utility function
called merge_dict which will recursively merge a dictionary so that individual
sub-components of the product catalog data can be updated without other
data being overwritten.

This commit also separates the util module into separate files, k8s.py
and merge_dict.py, to provide more separation between the merge_dict
function and its helpers and the load_k8s utility. To preserve the ability
to import load_k8s from cray_product_catalog.util,
cray_product_catalog.util.k8s.load_k8s is imported in the util package's
__init__ module.

Finally, this commit adds a change log entry.

Test Description:

I tested this in a local minikube environment in which I created the
services/cray-product-catalog config map manually.

Ran two cray-product-catalog-update processes (A and B), with a debug
break point in process A immediately prior to calling
replace_namespaced_config_map.

This resulted in process B updating the config map, without conflict,
between the time that process A read the config map and the time that
it attempted to replace it. As expected, this resulted in process A
receiving a conflict response and required it to re-read the config map
and update it once again. In the end, both process A and process B's
updates were reflected in the config map.

I tested inserting data into a large (i.e. nearly 1M) config map to
validate there were no issues with large config maps.

I also wrote unit tests for the merge_dict utility.

## Summary and Scope

See commit message

## Issues and Related PRs

_List and characterize relationship to Jira/Github issues and other pull requests. Be sure to list dependencies._

* Resolves CASM-3555 and CASM-3556

## Testing

See commit message

## Pull Request Checklist

- [X] Target branch correct
- [X] Testing is appropriate and complete, if applicable
- [X] Is a new version being released? Update https://github.com/Cray-HPE/cray-product-catalog/wiki/CSM-Compatibility-Matrix
- [X] [HPC Product Announcement](https://cray.slack.com/archives/C026TVCSXLH) prepared, if applicable

### CODE REVIEW

File: cray_product_catalog/util/merge_dict.py
Comment (jack-stanek-hpe) [inline cray_product_catalog/util/merge_dict.py:120]: nit: `TypeError` might be more appropriate here

**Review by jack-stanek-hpe — APPROVED** [review]
Looks good to me.

File: cray_product_catalog/util/merge_dict.py
Comment (ekamin-hpe) [reply cray_product_catalog/util/merge_dict.py]: Thanks, fixed in 24778dd5947ad8f48dbadf2ded6d403a2d3ca6b6

File: cray_product_catalog/util/merge_dict.py
Comment (ekamin-hpe) [reply cray_product_catalog/util/merge_dict.py]: Thanks, fixed in 24778dd5947ad8f48dbadf2ded6d403a2d3ca6b6

**Comment by ekamin-hpe** [comment]
Thank you for the review, @jack-stanek-hpe. This still needs a code owner review from a member of @Cray-HPE/cray-management-systems.

File: cray_product_catalog/util/merge_dict.py
Comment (jsollom-hpe) [inline cray_product_catalog/util/merge_dict.py:85]: The second isinstance is unnecessary.

File: tests/util/mocks.py
Comment (jsollom-hpe) [inline tests/util/mocks.py:94]: Yum!

**Review by jsollom-hpe — APPROVED** [review]

File: cray_product_catalog/util/merge_dict.py
Comment (ekamin-hpe) [reply cray_product_catalog/util/merge_dict.py:85]: I wanted to handle (admittedly pretty edge-like) cases where we might be dealing with a subclass of a dictionary.
```
>>> class MySpecialDict(dict):
...   pass
...
>>> dictA = MySpecialDict()
>>> dictB = {}
>>> isinstance(dictA, type(dictB))
True
>>> isinstance(dictB, type(dictA))
False
```

In this case I think I would like the utility to consider a `MySpecialDict` and a regular `dict` to be the same type regardless of the order of the arguments.

Comment (ekamin-hpe) [reply cray_product_catalog/util/merge_dict.py:85]: I'm going to resolve conversation this because at worst I think it is just a potentially unnecessary/overkill line of code.

Comment (ekamin-hpe) [reply cray_product_catalog/util/merge_dict.py:85]: Maybe we shouldn't allow subclasses to compare as the same. I think that would mean making the `or` into an `and`.

**Review by jack-stanek-hpe — DISMISSED** [review]

File: cray_product_catalog/util/merge_dict.py
Comment (ekamin-hpe) [reply cray_product_catalog/util/merge_dict.py:85]: Thinking about it some more, I'm actually going to go with that and `and` the two comparisons together. I don't think we should be allowing subclasses and their parent class to be considered the same type for the purposes of this utility. Realistically I don't see how this would matter if we're just dealing with parsed yaml structures, but I can't guarantee that this utility won't be used for something different in the future.

**Review by jsollom-hpe — APPROVED** [review]

File: cray_product_catalog/util/merge_dict.py
Comment (jack-stanek-hpe) [inline cray_product_catalog/util/merge_dict.py:86]: Based on your reasoning [here](https://github.com/Cray-HPE/cray-product-catalog/pull/160/files#r1019562916) I think this can just be simplified to the following?
```suggestion
    return type(first_value) != type(second_value)
```
This handles both subclasses and disparate types properly.

**Review by jack-stanek-hpe — APPROVED** [review]

File: cray_product_catalog/util/merge_dict.py
Comment (ekamin-hpe) [reply cray_product_catalog/util/merge_dict.py:86]: In that case I end up getting a linter warning about comparing two `type`s, if you're good with it I'm just going to leave this.

**Pull request merged**
